### ISSUE TITLE
Correction for $matrix evaluation

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#3423]
Imported from SourceForge on 2024-07-08 00:47:06
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2013-10-29 01:05:49
Original: https://sourceforge.net/p/maxima/bugs/2649

---

apply('matrix,'[[print(3)]]) prints 3 -- that is, it evaluates its argument twice. This is a bug that causes problems in, e.g., mat_function. See http://comments.gmane.org/gmane.comp.mathematics.maxima.general/43716

The fix is to redefine $matrix as a regular function (not a defmspec):

    (remprop '$matrix 'mfexpr*)
    (defun $matrix (&rest rows)
      (dolist (row rows)
        (if (not ($listp row)) (merror (intl:gettext "matrix: row must be a list; found: ~M") row)))
      (matcheck rows)
      (cons '($matrix) rows))

I have NOT run the test suite after adding this fix.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-08 00:47:08
Created by **[villate](https://sourceforge.net/u/villate/)** on 2013-11-01 08:42:35
Original: https://sourceforge.net/p/maxima/bugs/2649/#13cd

---

- Description has changed:

Diff:

~~~~

--- old
+++ new
@@ -2,11 +2,11 @@
 
 The fix is to redefine $matrix as a regular function (not a defmspec):
 
-(remprop '$matrix 'mfexpr*)
-(defun $matrix (&rest rows)
-  (dolist (row rows)
-    (if (not ($listp row)) (merror (intl:gettext "matrix: row must be a list; found: ~M") row)))
-  (matcheck rows)
-  (cons '($matrix) rows))
+    (remprop '$matrix 'mfexpr*)
+    (defun $matrix (&rest rows)
+      (dolist (row rows)
+        (if (not ($listp row)) (merror (intl:gettext "matrix: row must be a list; found: ~M") row)))
+      (matcheck rows)
+      (cons '($matrix) rows))
 
 I have NOT run the test suite after adding this fix.

~~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-08 00:47:11
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2013-11-20 20:52:35
Original: https://sourceforge.net/p/maxima/bugs/2649/#ed68

---

- **labels**:  --> Lisp Core
- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-08 00:47:15
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2013-11-20 20:52:35
Original: https://sourceforge.net/p/maxima/bugs/2649/#cd89

---

Fixed by commit 31c58ed22 (rework $MATRIX in a slightly different way than shown here).
